### PR TITLE
[Rest Server] Add secrets field in protocol

### PR DESCRIPTION
**Pull request by abuccts**
Add secrets field in protocol, implement #2729:

* Update protocol schema
* Parse and render secrets field in commands/auth
* Support Docker login using secrets, remove credentials from `set -x` debug log
* Mask secrets field when storing yaml

### CODE REVIEW

File: src/rest-server/src/models/v2/job.js [hunk 1]
```diff
@@ -232,9 +233,13 @@ const prepareContainerScripts = async (frameworkName, userName, config, rawConfi
     JSON.stringify(frameworkDescription, null, 2), userName, '644')
   );
   // upload original config file to hdfs
+  // mask secrets field before uploading
+  let maskRawConfig = rawConfig + '\nZ';
```
Comment (abuccts) [inline src/rest-server/src/models/v2/job.js]: Regex in JavaScript does not support perl regex `\Z`, so add a newline at the end manually.

File: src/rest-server/src/middlewares/v2/protocol.js [hunk 2]
```diff
@@ -107,6 +128,18 @@ const protocolValidate = (protocolYAML) => {
 };
 
 const protocolRender = (protocolObj) => {
+  // render auth for Docker image
+  for (let name of Object.keys(protocolObj.prerequisites.dockerimage)) {
+    if ('auth' in protocolObj.prerequisites.dockerimage[name]) {
+      for (let prop of Object.keys(protocolObj.prerequisites.dockerimage[name].auth)) {
+        protocolObj.prerequisites.dockerimage[name].auth[prop] = render(
+          protocolObj.prerequisites.dockerimage[name].auth[prop],
+          {'$secrets': protocolObj.secrets},
```
Comment (fanyangCS) [inline src/rest-server/src/middlewares/v2/protocol.js]: you assume the auth field must use secrets?

File: src/rest-server/src/models/v2/job.js (hunk 1 above)
Comment (fanyangCS) [reply src/rest-server/src/models/v2/job.js]: ut please.

---
In reply to: [283308207](https://github.com/microsoft/pai/pull/2752#discussion_r283308207) [](ancestors = 283308207)

File: src/rest-server/src/middlewares/v2/protocol.js [hunk 3]
```diff
@@ -39,6 +39,27 @@ const prerequisiteFields = [
   'dockerImage',
 ];
 
+
+const render = (template, dict, tags=['<%', '%>']) => {
```
[the remark this answers is not in the document]
Comment (abuccts) [reply src/rest-server/src/middlewares/v2/protocol.js]: this function has already been covered in unit test, added test case for yaml with secrets field

File: src/rest-server/src/middlewares/v2/protocol.js [hunk 4]
```diff
@@ -107,6 +128,18 @@ const protocolValidate = (protocolYAML) => {
 };
 
 const protocolRender = (protocolObj) => {
+  // render auth for Docker image
```
[the remark this answers is not in the document]
Comment (abuccts) [reply src/rest-server/src/middlewares/v2/protocol.js]: this function has already been covered in unit test, added more test cases

File: src/rest-server/src/middlewares/v2/protocol.js (hunk 2 above)
Comment (abuccts) [reply src/rest-server/src/middlewares/v2/protocol.js]: no, the value can be null or undefined

File: src/rest-server/src/models/v2/job.js (hunk 1 above)
Comment (abuccts) [reply src/rest-server/src/models/v2/job.js]: split into a separate function and added unit test

**Pull request merged**